[Bug]: Some tests are failing due to ImportError within schemasheets

### What happened?

These tests are failing:

![Image](https://github.com/user-attachments/assets/eea45698-842b-4ee9-b6f2-2916914fbb3a)

Because of an incompatibility between schemasheets and linkml-runtime after an update to linkml-runtime:

```
E   ImportError: cannot import name 're' from 'linkml_runtime.utils.schemaview' (/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/linkml_runtime/utils/schemaview.py)
```

Because this is an ImportError, HDMF incorrectly thinks that schemasheets is not installed and so it raises the error:
```
E           ValueError: Install schemasheets.
```

I have created PRs to fix these issues in schemasheets: https://github.com/linkml/schemasheets/pull/148 and https://github.com/linkml/schemasheets/pull/149 that are awaiting review by the linkml team

If that takes too long to resolve, I suggest we limit linkml-runtime < 1.8.0 (this prevents us from using python 3.13 with linkml/schemasheets but that's OK).

### Steps to Reproduce

```python
Install and run "from schemasheets.schemamaker import SchemaMaker"
```

### Traceback

```pytb

```

### Operating System

macOS

### Python Version

3.12

### Package Versions

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,12 +4,14 @@
 
 ### Enhancements
 - Optimized `get` within `VectorIndex` to be more efficient when retrieving a dataset of references. @mavaylon1 [#1248](https://github.com/hdmf-dev/hdmf/pull/1248)
+- Improved error messages when optional requirements are not installed. @rly [#1263](https://github.com/hdmf-dev/hdmf/pull/1263)
 
 ### Changed
 - Removed `requirements-min.txt` in favor of the `min-reqs` optional dependency group in `pyproject.toml`. @rly [#1246](https://github.com/hdmf-dev/hdmf/pull/1246)
 - Updated GitHub actions and ruff configuration. @rly [#1246](https://github.com/hdmf-dev/hdmf/pull/1246)
 - `hdmf.monitor` is unused and undocumented. It has been deprecated and will be removed in HDMF 5.0. @rly [#1245](https://github.com/hdmf-dev/hdmf/pull/1245)
 - Restricted numcodecs dependency to <0.16 due to incompatibility with the latest zarr<3 version. @rly [#1257](https://github.com/hdmf-dev/hdmf/pull/1257)
+- Increased minimum requirement for optional schemasheets dependency to 0.4.0rc1 due to incompatibility with linkml-runtime. @rly [#1263](https://github.com/hdmf-dev/hdmf/pull/1263)
 
 ## HDMF 4.0.0 (January 22, 2025)
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -49,7 +49,7 @@ zarr = [
 sparse = ["scipy>=1.7"]
 termset = [
     "linkml-runtime>=1.5.5",
-    "schemasheets>=0.1.23",
+    "schemasheets>=0.4.0rc1",
     "oaklib>=0.5.12",
     "pyyaml>=6.0.1",
 ]

--- a/src/hdmf/term_set.py
+++ b/src/hdmf/term_set.py
@@ -33,7 +33,10 @@ class TermSet:
         try:
             from linkml_runtime.utils.schemaview import SchemaView
         except ImportError:
-            msg = "There is an issue with importing linkml_runtime. Please make sure a compatible version of linkml_runtime is installed."
+            msg = (
+                "There is an issue with importing linkml_runtime. Please make sure a "
+                "compatible version of linkml_runtime is installed."
+            )
             raise ValueError(msg)
 
         self.term_schema_path = term_schema_path
@@ -166,7 +169,10 @@ class TermSet:
             from linkml_runtime.utils.schema_as_dict import schema_as_dict
             from schemasheets.schemamaker import SchemaMaker
         except ImportError as e:   # pragma: no cover
-            msg = "There is an issue with importing schemasheets. Please make sure a compatible version of schemascheets is installed."
+            msg = (
+                "There is an issue with importing schemasheets. Please make sure a compatible "
+                "version of schemascheets is installed."
+            )
             raise ValueError(msg) from e
 
         schema_maker = SchemaMaker()
@@ -194,7 +200,10 @@ class TermSet:
                 warnings.filterwarnings("ignore", category=DeprecationWarning)
                 from oaklib.utilities.subsets.value_set_expander import ValueSetExpander
         except ImportError:   # pragma: no cover
-            msg = "There is an issue with importing oaklib. Please make sure a compatible version of oaklib is installed."
+            msg = (
+                "There is an issue with importing oaklib. Please make sure a compatible "
+                "version of oaklib is installed."
+            )
             raise ValueError(msg)
         expander = ValueSetExpander()
         # TODO: linkml should raise a warning if the schema does not have dynamic enums

--- a/src/hdmf/term_set.py
+++ b/src/hdmf/term_set.py
@@ -37,7 +37,7 @@ class TermSet:
                 "There is an issue with importing linkml_runtime. Please make sure a "
                 "compatible version of linkml_runtime is installed."
             )
-            raise ValueError(msg)
+            raise ImportError(msg)
 
         self.term_schema_path = term_schema_path
         self.schemasheets_folder = schemasheets_folder
@@ -173,7 +173,7 @@ class TermSet:
                 "There is an issue with importing schemasheets. Please make sure a compatible "
                 "version of schemascheets is installed."
             )
-            raise ValueError(msg) from e
+            raise ImportError(msg) from e
 
         schema_maker = SchemaMaker()
         tsv_file_paths = glob.glob(self.schemasheets_folder + "/*.tsv")
@@ -204,7 +204,7 @@ class TermSet:
                 "There is an issue with importing oaklib. Please make sure a compatible "
                 "version of oaklib is installed."
             )
-            raise ValueError(msg)
+            raise ImportError(msg)
         expander = ValueSetExpander()
         # TODO: linkml should raise a warning if the schema does not have dynamic enums
         enum = list(self.view.all_enums())

--- a/src/hdmf/term_set.py
+++ b/src/hdmf/term_set.py
@@ -33,7 +33,7 @@ class TermSet:
         try:
             from linkml_runtime.utils.schemaview import SchemaView
         except ImportError:
-            msg = "Install linkml_runtime"
+            msg = "There is an issue with importing linkml_runtime. Please make sure a compatible version of linkml_runtime is installed."
             raise ValueError(msg)
 
         self.term_schema_path = term_schema_path
@@ -165,9 +165,9 @@ class TermSet:
         try:
             from linkml_runtime.utils.schema_as_dict import schema_as_dict
             from schemasheets.schemamaker import SchemaMaker
-        except ImportError:   # pragma: no cover
-            msg = "Install schemasheets."
-            raise ValueError(msg)
+        except ImportError as e:   # pragma: no cover
+            msg = "There is an issue with importing schemasheets. Please make sure a compatible version of schemascheets is installed."
+            raise ValueError(msg) from e
 
         schema_maker = SchemaMaker()
         tsv_file_paths = glob.glob(self.schemasheets_folder + "/*.tsv")
@@ -194,7 +194,7 @@ class TermSet:
                 warnings.filterwarnings("ignore", category=DeprecationWarning)
                 from oaklib.utilities.subsets.value_set_expander import ValueSetExpander
         except ImportError:   # pragma: no cover
-            msg = 'Install oaklib.'
+            msg = "There is an issue with importing oaklib. Please make sure a compatible version of oaklib is installed."
             raise ValueError(msg)
         expander = ValueSetExpander()
         # TODO: linkml should raise a warning if the schema does not have dynamic enums

--- a/tests/unit/test_term_set.py
+++ b/tests/unit/test_term_set.py
@@ -351,3 +351,22 @@ class TestGlobalTypeConfig(TestCase):
             ExtensionContainer(name='foo',
                                namespace='foo_namespace',
                                description='Homo sapiens')
+
+
+class TestOptionalDepsNotInstalled(TestCase):
+
+    def setUp(self):
+        if REQUIREMENTS_INSTALLED:
+            self.skipTest("optional modules are installed")
+
+    def test_schemasheets_not_installed(self):
+        with self.assertRaises(ImportError):
+            TermSet(schemasheets_folder="tests/unit/test_term_set_input/schemasheets")
+
+    def test_linkml_runtime_not_installed(self):
+        with self.assertRaises(ImportError):
+            TermSet(term_schema_path="tests/unit/example_test_term_set.yaml")
+
+    def test_oaklib_not_installed(self):
+        with self.assertRaises(ImportError):
+            TermSet(term_schema_path="tests/unit/example_dynamic_term_set.yaml", dynamic=True)

